psi4/driver/qcdb/molparse/from_string.py

```diff
@@ -0,0 +1,721 @@
+
+    Raises
+    ------
+    MoleculeFormatError
+        After processing of `molstr`, only an empty string should remain.
+        Anything left is a syntax error.
+
+    Notes
+    -----
+        xyz - Strict XYZ format
+        -----------------------
+
+            String Layout
+            -------------
+            <number of atoms>
+            comment line
+            <element_symbol or atomic_number> <x> <y> <z>
+            ...
+            <element_symbol or atomic_number> <x> <y> <z>
+
+            QM Domain
+            ---------
+            Specifiable: geom, elem/elez (element identity)
+            Inaccessible: mass, real (vs. ghost), elbl (user label), name, units (assumed [A]),
+                          input_units_to_au, fix_com/orientation/symmetry, fragmentation,
+                          molecular_charge, molecular_multiplicity
+
+            Notes
+            -----
+            <number of atoms> is pattern-matched but ignored.
+
+        xyz+ - Enhanced XYZ format
+        --------------------------
+
+            String Layout
+            -------------
+            <number of atoms> [<bohr|au|ang>]
+            [<molecular_charge> <molecular_multiplicity>] comment line
+            <psi4_nucleus_spec> <x> <y> <z>
+            ...
+            <psi4_nucleus_spec> <x> <y> <z>
+
+            QM Domain
+            ---------
+            Specifiable: geom, elem/elez (element identity), mass, real (vs. ghost), elbl (user label),
+                         units (defaults [A]), molecular_charge, molecular_multiplicity
+            Inaccessible: name, input_units_to_au, fix_com/orientation/symmetry, fragmentation
+
+            Notes
+            -----
+            <number of atoms> is pattern-matched but ignored.
+
+        psi4 - Psi4 molecule {...} format
+        ---------------------------------
+
+            QM Domain
+            ---------
+            Specifiable: geom, elem/elez (element identity), mass, real (vs. ghost), elbl (user label),
+                         units (defaults [A]), fix_com/orientation/symmetry, fragment_separators,
+                         fragment_charges, fragment_multiplicities, molecular_charge, molecular_multiplicity
+            Inaccessible: name, input_units_to_au
+
+            EFP Domain
+            ----------
+            Specifiable: units, fix_com/orientation/symmetry, fragment_files, hint_types, geom_hints
+            Inaccessible: anything atomic or fragment details -- geom, elem/elez (element identity),
+                          mass, real (vs. ghost), elbl (user label), fragment_separators, fragment_charges,
+                          fragment_multiplicities, molecular_charge, molecular_multiplicity
+
+        psi4+ - Psi4 non-Cartesian molecule {...} format
+        ------------------------------------------------
+        Like `dtype=psi4` (although combination with EFP not tested) except
+        that instead of pure-Cartesian geometry, allow variables, zmatrix,
```

Comment:
I forget if we have documentation of the fact that zmatrices are not really used for optimization purposes.  Might be worth noting that here.

Reply:
Huh, apparently we did (".. warning:: The selection of a Z-matrix input, and in particular the inclusion
   of dummy atoms, has no effect on the behavior of the optimizer, which begins
   from a Cartesian representation of the system."), not that I remembered it.

I added it here, too, for good measure. Good idea — it needs drumming in.